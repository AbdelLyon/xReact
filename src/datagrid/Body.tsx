import { memo, useCallback } from "react";

import {
  AccordionItem,
  AccordionTrigger,
  AccordionContent,
} from "@/ui/accordion";
import { Checkbox } from "@/input";
import { cn } from "@/utils";
import { BodyProps } from "@/types/dataGridTypes";

export interface ColumnDefinition<T> {
  field?: keyof T;
  header: string;
  className?: string;
  cell?: (row: T, column: ColumnDefinition<T>) => React.ReactNode;
}

export const Body = memo(
  <T extends { id: string | number }>({
    row,
    columns,
    isChecked,
    onCheckboxChange,
    children,
    observerRef,
    checkboxSelection,
    onAccordionClick,
    accordionContent,
  }: React.PropsWithChildren<BodyProps<T>>) => {
    const renderColumns = (
      column: ColumnDefinition<T>,
      row: T,
    ): React.ReactNode => {
      if (column.cell) {
        return column.cell(row, column);
      }

      const cellValue =
        column.field !== undefined ? row[column.field] : undefined;

      return cellValue as React.ReactNode;
    };

    const handleClick = useCallback(() => {
      onAccordionClick?.(row);
    }, [onAccordionClick, row]);

    return (
      <AccordionItem
        ref={observerRef}
        className="w-full last:border-b-0"
        value={String(row.id)}
      >
        <div className="relative flex h-14 w-full items-center px-4">
          {checkboxSelection && (
            <Checkbox
              checked={isChecked}
              className="mr-2"
              onCheckedChange={() => onCheckboxChange(row)}
            />
          )}
          {columns.map((column) => (
            <div
              key={`cell-${row.id}-${String(column.field)}`}
              className={cn("flex-1 text-left", column.className)}
            >
              {renderColumns(column as ColumnDefinition<T>, row)}
            </div>
          ))}
          {(accordionContent || children) && (
            <AccordionTrigger
              className="absolute right-4 top-1/2 flex size-6 -translate-y-1/2 items-center justify-center"
              onClick={handleClick}
            />
          )}
        </div>
        {(accordionContent || children) && (
          <AccordionContent className="rounded-t px-2 shadow">
            {accordionContent || children}
          </AccordionContent>
        )}
      </AccordionItem>
    );
  },
);

Body.displayName = "Body";
