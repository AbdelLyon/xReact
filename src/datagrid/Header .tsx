// Header.tsx
import { ChevronDown, ChevronUp } from "lucide-react";
import { memo } from "react";

import { AccordionItem } from "@/ui/accordion";
import { Checkbox } from "@/input";
import { cn } from "@/utils";
import { HeaderProps } from "@/types/dataGridTypes";

export const Header = memo(
  <T extends { id: string | number }>({
    columns,
    isAllChecked,
    onSelectAll,
    checkboxSelection,
    sortConfig,
    onSort,
  }: HeaderProps<T>) => (
    <AccordionItem
      className="bg-card-light flex h-14 w-full items-center rounded-t-md px-4 py-3 dark:bg-card"
      value="header"
    >
      <div className="flex h-14 w-full items-center">
        {checkboxSelection && (
          <Checkbox
            checked={isAllChecked}
            className="mr-2 dark:border-white/5"
            onCheckedChange={onSelectAll}
          />
        )}
        {columns.map((column) => (
          <button
            key={`header-${String(column.field)}`}
            className={cn(
              "flex-1 font-semibold text-left flex items-center justify-between gap-2",
              column.className,
            )}
            onClick={() =>
              column.sortable &&
              column.field &&
              onSort(
                column.field,
                sortConfig.direction === "asc" ? "desc" : "asc",
              )
            }
          >
            <div className="flex items-center gap-1">
              <p className="grow opacity-80">{column.header}</p>
              {column.sortable && column.field && (
                <div className="relative size-4 shrink-0">
                  <ChevronUp
                    className={cn(
                      "absolute -top-1 transition-opacity duration-200",
                      sortConfig.key === column.field &&
                        sortConfig.direction === "asc"
                        ? "opacity-80"
                        : "opacity-30",
                    )}
                    size={15}
                  />
                  <ChevronDown
                    className={cn(
                      "absolute top-1 transition-opacity duration-200",
                      sortConfig.key === column.field &&
                        sortConfig.direction === "desc"
                        ? "opacity-80"
                        : "opacity-30",
                    )}
                    size={15}
                  />
                </div>
              )}
            </div>
          </button>
        ))}
      </div>
    </AccordionItem>
  ),
);

Header.displayName = "Header";
