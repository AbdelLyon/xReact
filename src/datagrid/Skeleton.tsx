import { TableBody, TableCell, TableRow } from "@/ui/table";
import { cn } from "@/utils";
import { AccordionContent, AccordionItem } from "@/ui/accordion";
import { ColumnDefinition } from "@/types/dataGridTypes";

type SkeletonRowsProps<T> = {
  columns: ColumnDefinition<T>[];
  skeletonRows: number;
};

export function SkeletonRows<T>({
  columns,
  skeletonRows,
}: Readonly<SkeletonRowsProps<T>>) {
  return (
    <TableBody>
      {Array(skeletonRows)
        .fill(null)
        .map((_, rowIndex) => (
          <TableRow key={`skeleton-row-${rowIndex}`}>
            <TableCell className="w-[50px]">
              <div className="size-4 animate-pulse rounded bg-gray-200 dark:bg-gray-800" />
            </TableCell>
            {columns.map((column, cellIndex) => (
              <TableCell
                key={`skeleton-cell-${rowIndex}-${cellIndex}`}
                className={column.className}
              >
                <div className="h-4 animate-pulse rounded bg-gray-200 dark:bg-gray-800" />
              </TableCell>
            ))}
            <TableCell className="w-[50px]">
              <div className="size-4 animate-pulse rounded bg-gray-200 dark:bg-gray-800" />
            </TableCell>
          </TableRow>
        ))}
    </TableBody>
  );
}

export const SkeletonRowAccordion = <T extends { id: string | number }>({
  columns,
  index,
}: {
  columns: ColumnDefinition<T>[];
  index: number;
}) => (
  <AccordionItem
    key={`skeleton-${index}`}
    className="w-full border-b last:border-b-0"
    value={`skeleton-${index}`}
  >
    <div className="flex h-14 w-full items-center px-4">
      <div className="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card" />{" "}
      {columns.map((column, colIndex) => (
        <div
          key={`skeleton-cell-${index}-${colIndex}`}
          className={cn("flex-1", column.className)}
        >
          <div className="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card" />
        </div>
      ))}
      <div className="size-4 animate-pulse rounded-sm bg-border dark:bg-card" />
    </div>
    <AccordionContent className="h-4 w-full bg-border dark:bg-card">
      <div className="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card" />
    </AccordionContent>
  </AccordionItem>
);
