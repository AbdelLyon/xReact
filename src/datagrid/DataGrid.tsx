import { useMemo } from "react";
import { useInView } from "react-intersection-observer";

import { SkeletonRowAccordion } from "./Skeleton";
import { Body } from "./Body";
import { Header } from "./Header ";

import { Accordion } from "@/ui/accordion";
import { cn } from "@/utils";
import { useInfiniteScroll } from "@/hooks";
import { LoadingSpinner } from "@/shared/input/LoadingSpinner";
import { useDataGridState } from "@/hooks/useDataGridState";
import {
  ColumnDefinition,
  DataGridProps,
  SortConfig,
} from "@/types/dataGridTypes";

export function DataGrid<T extends { id: string | number }>({
  rows,
  columns,
  caption,
  className,
  footerContent,
  isLoading,
  skeletonRows = 10,
  onCheckedRowsChange,
  children,
  onFetchPage,
  isFetching,
  checkboxSelection = true,
  onSort,
  onAccordionClick,
  accordionContent,
}: React.PropsWithChildren<Readonly<DataGridProps<T>>>) {
  const { ref, inView } = useInView();

  useInfiniteScroll(inView, onFetchPage);

  const {
    checkedRows,
    isAllChecked,
    sortConfig,
    handleCheckboxChange,
    handleSelectAll,
    handleSort,
  } = useDataGridState<T>(rows, onCheckedRowsChange, onSort);

  const memoizedColumns = useMemo(() => columns, [columns]);

  const renderSkeletons = useMemo(
    () =>
      Array(skeletonRows)
        .fill(null)
        .map((_, index) => (
          <SkeletonRowAccordion
            key={`skeleton-row-${crypto.randomUUID()}`}
            columns={memoizedColumns}
            index={index}
          />
        )),
    [skeletonRows, memoizedColumns],
  );

  const renderRows = useMemo(
    () =>
      rows.map((row, i: number) => {
        const expandedNode = accordionContent ? accordionContent(row) : null;

        return (
          <Body
            key={row.id}
            accordionContent={
              expandedNode as
                | ((row: { id: string | number }) => React.ReactNode)
                | undefined
            }
            checkboxSelection={checkboxSelection}
            columns={memoizedColumns as ColumnDefinition<any>[]}
            i={i}
            isChecked={checkedRows.has(row)}
            observerRef={i === rows.length - 1 ? ref : undefined}
            row={row}
            rows={rows}
            onAccordionClick={
              onAccordionClick as
                | ((row: { id: string | number }) => void)
                | undefined
            }
            onCheckboxChange={() => handleCheckboxChange(row)}
          >
            {expandedNode || children}
          </Body>
        );
      }),
    [
      rows,
      memoizedColumns,
      checkedRows,
      handleCheckboxChange,
      ref,
      checkboxSelection,
      onAccordionClick,
      accordionContent,
    ],
  );

  return (
    <div
      className={cn(
        "bg-white dark:bg-background border border-border relative rounded-t-md shadow-lg overflow-auto",
        className,
      )}
    >
      {caption && (
        <div className="p-4 text-center text-lg font-semibold">{caption}</div>
      )}
      <Accordion collapsible className="relative w-full" type="single">
        <Header
          checkboxSelection={checkboxSelection}
          columns={memoizedColumns as ColumnDefinition<any>[]}
          isAllChecked={isAllChecked}
          sortConfig={sortConfig as SortConfig}
          onSelectAll={handleSelectAll}
          onSort={handleSort}
        />

        {isLoading ? renderSkeletons : renderRows}
        {isFetching && <LoadingSpinner />}
      </Accordion>
      {footerContent && (
        <div className="border-t border-border p-4">{footerContent}</div>
      )}
    </div>
  );
}
