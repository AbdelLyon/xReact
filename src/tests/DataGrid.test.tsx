import { describe, it, expect, vi } from "vitest";
import { render } from "@testing-library/react";

import { ColumnDefinition, DataGrid } from "@/datagrid";

type Columns = ColumnDefinition<{
  id: string | number;
}>;

vi.mock("@/ui/accordion", () => ({
  Accordion: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="accordion">{children}</div>
  ),
  AccordionItem: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="accordion-item">{children}</div>
  ),
  AccordionTrigger: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="accordion-trigger">{children}</div>
  ),
  AccordionContent: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="accordion-content">{children}</div>
  ),
}));

vi.mock("@/hooks", () => ({
  useInfiniteScroll: vi.fn(),
}));

vi.mock("react-intersection-observer", () => ({
  useInView: () => ({ ref: vi.fn(), inView: false }),
}));

vi.mock("@/hooks/useDataGridState", () => ({
  useDataGridState: () => ({
    checkedRows: new Set(),
    isAllChecked: false,
    sortConfig: { key: null, direction: "asc" },
    handleCheckboxChange: vi.fn(),
    handleSelectAll: vi.fn(),
    handleSort: vi.fn(),
  }),
}));

vi.mock("./Skeleton", () => ({
  SkeletonRowAccordion: () => <div data-testid="skeleton-row" />,
}));

vi.mock("./Body", () => ({
  Body: () => <div data-testid="body-row" />,
}));

vi.mock("./Header", () => ({
  Header: ({ columns, isAllChecked, checkboxSelection }: any) => (
    <div data-testid="header">
      <span>Columns: {columns.length}</span>
      <span>Is All Checked: {isAllChecked.toString()}</span>
      <span>Checkbox Selection: {checkboxSelection.toString()}</span>
    </div>
  ),
}));

vi.mock("@/shared/input/LoadingSpinner", () => ({
  LoadingSpinner: () => <div data-testid="loading-spinner" />,
}));

vi.mock("@/utils", () => ({
  cn: (...args: any[]) => args.join(" "),
}));

// Tests
describe("DataGrid", () => {
  const mockColumns = [
    { field: "id", header: "ID" },
    { field: "name", header: "Name" },
  ];

  const mockRows = [
    { id: 1, name: "John Doe" },
    { id: 2, name: "Jane Doe" },
  ];

  it("renders correctly with basic props", () => {
    const { container } = render(
      <DataGrid columns={mockColumns as Columns[]} rows={mockRows} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders loading state correctly", () => {
    const { container } = render(
      <DataGrid
        columns={mockColumns as Columns[]}
        isLoading={true}
        rows={[]}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders with caption and footer", () => {
    const { container } = render(
      <DataGrid
        caption="Test Caption"
        columns={mockColumns as Columns[]}
        footerContent={<div>Footer Content</div>}
        rows={mockRows}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders with fetching state", () => {
    const { container } = render(
      <DataGrid
        columns={mockColumns as Columns[]}
        isFetching={true}
        rows={mockRows}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
