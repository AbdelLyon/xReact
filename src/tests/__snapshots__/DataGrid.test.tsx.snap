// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataGrid > renders correctly with basic props 1`] = `
<div>
  <div
    class="bg-white dark:bg-background border border-border relative rounded-t-md shadow-lg overflow-auto "
  >
    <div
      data-testid="accordion"
    >
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2 dark:border-white/5"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                ID
              </p>
            </div>
          </button>
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                Name
              </p>
            </div>
          </button>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="relative flex h-14 w-full items-center px-4"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <div
            class="flex-1 text-left "
          >
            1
          </div>
          <div
            class="flex-1 text-left "
          >
            John Doe
          </div>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="relative flex h-14 w-full items-center px-4"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <div
            class="flex-1 text-left "
          >
            2
          </div>
          <div
            class="flex-1 text-left "
          >
            Jane Doe
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataGrid > renders loading state correctly 1`] = `
<div>
  <div
    class="bg-white dark:bg-background border border-border relative rounded-t-md shadow-lg overflow-auto "
  >
    <div
      data-testid="accordion"
    >
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2 dark:border-white/5"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                ID
              </p>
            </div>
          </button>
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                Name
              </p>
            </div>
          </button>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center px-4"
        >
          <div
            class="mr-2 size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
           
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="flex-1 "
          >
            <div
              class="h-4 w-3/4 animate-pulse rounded bg-border dark:bg-card"
            />
          </div>
          <div
            class="size-4 animate-pulse rounded-sm bg-border dark:bg-card"
          />
        </div>
        <div
          data-testid="accordion-content"
        >
          <div
            class="mt-2 h-4 w-1/2 animate-pulse rounded bg-border dark:bg-card"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DataGrid > renders with caption and footer 1`] = `
<div>
  <div
    class="bg-white dark:bg-background border border-border relative rounded-t-md shadow-lg overflow-auto "
  >
    <div
      class="p-4 text-center text-lg font-semibold"
    >
      Test Caption
    </div>
    <div
      data-testid="accordion"
    >
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2 dark:border-white/5"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                ID
              </p>
            </div>
          </button>
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                Name
              </p>
            </div>
          </button>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="relative flex h-14 w-full items-center px-4"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <div
            class="flex-1 text-left "
          >
            1
          </div>
          <div
            class="flex-1 text-left "
          >
            John Doe
          </div>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="relative flex h-14 w-full items-center px-4"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <div
            class="flex-1 text-left "
          >
            2
          </div>
          <div
            class="flex-1 text-left "
          >
            Jane Doe
          </div>
        </div>
      </div>
    </div>
    <div
      class="border-border border-t p-4"
    >
      <div>
        Footer Content
      </div>
    </div>
  </div>
</div>
`;

exports[`DataGrid > renders with fetching state 1`] = `
<div>
  <div
    class="bg-white dark:bg-background border border-border relative rounded-t-md shadow-lg overflow-auto "
  >
    <div
      data-testid="accordion"
    >
      <div
        data-testid="accordion-item"
      >
        <div
          class="flex h-14 w-full items-center"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2 dark:border-white/5"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                ID
              </p>
            </div>
          </button>
          <button
            class="flex-1 font-semibold text-left flex items-center justify-between gap-2 "
          >
            <div
              class="flex items-center gap-1"
            >
              <p
                class="grow opacity-80"
              >
                Name
              </p>
            </div>
          </button>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="relative flex h-14 w-full items-center px-4"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <div
            class="flex-1 text-left "
          >
            1
          </div>
          <div
            class="flex-1 text-left "
          >
            John Doe
          </div>
        </div>
      </div>
      <div
        data-testid="accordion-item"
      >
        <div
          class="relative flex h-14 w-full items-center px-4"
        >
          <button
            aria-checked="false"
            class="peer size-[18px] shrink-0 rounded-sm border border-border ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mr-2"
            data-state="unchecked"
            role="checkbox"
            type="button"
            value="on"
          />
          <div
            class="flex-1 text-left "
          >
            2
          </div>
          <div
            class="flex-1 text-left "
          >
            Jane Doe
          </div>
        </div>
      </div>
      <div
        data-testid="loading-spinner"
      />
    </div>
  </div>
</div>
`;
